Description: run tests in verbose mode
Author: Lucas Nussbaum <lucas@debian.org>

--- a/common.mk
+++ b/common.mk
@@ -492,15 +492,15 @@ test: test-sample btest-ruby test-knownb
 
 test-all: $(TEST_RUNNABLE)-test-all
 yes-test-all: PHONY
-	$(RUNRUBY) "$(srcdir)/test/runner.rb" --ruby="$(RUNRUBY)" $(TESTS)
+	$(RUNRUBY) "$(srcdir)/test/runner.rb" -v --ruby="$(RUNRUBY)" $(TESTS)
 TESTS_BUILD = mkmf
 no-test-all: PHONY
-	$(MINIRUBY) -I"$(srcdir)/lib" "$(srcdir)/test/runner.rb" $(TESTS_BUILD)
+	$(MINIRUBY) -I"$(srcdir)/lib" "$(srcdir)/test/runner.rb" -v $(TESTS_BUILD)
 
 test-ruby: $(TEST_RUNNABLE)-test-ruby
 no-test-ruby: PHONY
 yes-test-ruby: PHONY
-	$(RUNRUBY) "$(srcdir)/test/runner.rb" ruby
+	$(RUNRUBY) "$(srcdir)/test/runner.rb" -v ruby
 
 extconf: $(PREP)
 	$(Q) $(MAKEDIRS) "$(EXTCONFDIR)"
