--- a/lib/rubygems/defaults.rb
+++ b/lib/rubygems/defaults.rb
@@ -15,8 +15,13 @@ module Gem
   ##
   # Default home directory path to be used if an alternate value is not
   # specified in the environment
+  # 
+  # Debian patch: 
+  #   /var/lib/gems/{ruby version} (This is the default path in Debian system)
+  #
 
   def self.default_dir
+    return File.join('/', 'var', 'lib', 'gems', ConfigMap[:ruby_version])
     if defined? RUBY_FRAMEWORK_VERSION then
       File.join File.dirname(ConfigMap[:sitedir]), 'Gems',
                 ConfigMap[:ruby_version]
@@ -63,8 +68,12 @@ module Gem
 
   ##
   # The default directory for binaries
+  #
+  # Debian patch: 
+  #   /var/lib/gems/{ruby version}/bin is the default path in Debian system
 
   def self.default_bindir
+    return File.join('/', 'usr', 'local', 'bin')
     if defined? RUBY_FRAMEWORK_VERSION then # mac framework support
       '/usr/bin'
     else # generic install
